File: mobile/lib/models/sensor_reading.dart
class SensorReading {
  final String id;
  final String farmId;
  final double? soilMoisturePercent;
  final double? soilTemperatureCelsius;
  final double? soilPh;
  final double? nitrogenPpm;
  final double? phosphorusPpm;
  final double? potassiumPpm;
  final double? airTemperatureCelsius;
  final double? airHumidityPercent;
  final String recordedAt;

  const SensorReading({
    required this.id,
    required this.farmId,
    required this.recordedAt,
    this.soilMoisturePercent,
    this.soilTemperatureCelsius,
    this.soilPh,
    this.nitrogenPpm,
    this.phosphorusPpm,
    this.potassiumPpm,
    this.airTemperatureCelsius,
    this.airHumidityPercent,
  });

  factory SensorReading.fromJson(Map<String, dynamic> json) {
    return SensorReading(
      id: json['id']?.toString() ?? '',
      farmId: json['farm_id']?.toString() ?? '',
      soilMoisturePercent: (json['soil_moisture_percent'] as num?)?.toDouble(),
      soilTemperatureCelsius: (json['soil_temperature_celsius'] as num?)?.toDouble(),
      soilPh: (json['soil_ph'] as num?)?.toDouble(),
      nitrogenPpm: (json['nitrogen_ppm'] as num?)?.toDouble(),
      phosphorusPpm: (json['phosphorus_ppm'] as num?)?.toDouble(),
      potassiumPpm: (json['potassium_ppm'] as num?)?.toDouble(),
      airTemperatureCelsius: (json['air_temperature_celsius'] as num?)?.toDouble(),
      airHumidityPercent: (json['air_humidity_percent'] as num?)?.toDouble(),
      recordedAt: json['recorded_at']?.toString() ?? '',
    );
  }
}

class LatestSensorData {
  final String farmId;
  final SensorReading? latestReading;
  final String soilHealthStatus;
  final bool irrigationNeeded;
  final String? npkAlert;

  const LatestSensorData({
    required this.farmId,
    required this.soilHealthStatus,
    required this.irrigationNeeded,
    this.latestReading,
    this.npkAlert,
  });

  factory LatestSensorData.fromJson(Map<String, dynamic> json) {
    return LatestSensorData(
      farmId: json['farm_id']?.toString() ?? '',
      latestReading: json['latest_reading'] != null
          ? SensorReading.fromJson(json['latest_reading'] as Map<String, dynamic>)
          : null,
      soilHealthStatus: json['soil_health_status']?.toString() ?? 'unknown',
      irrigationNeeded: json['irrigation_needed'] == true,
      npkAlert: json['npk_alert']?.toString(),
    );
  }
}

File: backend/app/models/sensor_reading.py
from sqlalchemy import Column, Float, ForeignKey, DateTime, String
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func
import uuid
from app.database import Base


class SensorReading(Base):
    __tablename__ = "sensor_readings"

    id = Column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    farm_id = Column(String(36), ForeignKey("farms.id"), nullable=False, index=True)
    device_id = Column(String(100), nullable=False, index=True)
    soil_moisture_percent = Column(Float, nullable=True)
    soil_temperature_celsius = Column(Float, nullable=True)
    soil_ph = Column(Float, nullable=True)
    nitrogen_ppm = Column(Float, nullable=True)
    phosphorus_ppm = Column(Float, nullable=True)
    potassium_ppm = Column(Float, nullable=True)
    air_temperature_celsius = Column(Float, nullable=True)
    air_humidity_percent = Column(Float, nullable=True)
    light_lux = Column(Float, nullable=True)
    recorded_at = Column(DateTime(timezone=True), server_default=func.now(), index=True)

    farm = relationship("Farm", back_populates="sensor_readings")
